### DISCUSSION TITLE
hardhat-ethers: how to get RPC URL of ethers.provider?

### DISCUSSION THREAD

**Discussion by aspiers** [NomicFoundation/hardhat#2309] [Troubleshooting]
When using hardhat-ethers, `hre.ethers` is extended, and in particular `hre.ethers.provider` is assigned to a provider proxy which (IIUC) wraps around the underlying Hardhat provider.  However, the behaviour of this provider is not as transparent as it would be ideally.  In particular, `hre.ethers.provider.connection` seems to always return `{ url: 'http://localhost:8545' }` regardless of what RPC URL is actually being used.

I poked inside the underlying provider (`hre.ethers.provider._hardhatProvider`) and eventually managed to find the relevant connection info, but that was buried under an incredible number of layers of inheritance:

![image](https://user-images.githubusercontent.com/100738/151431099-56740d04-8063-4825-882c-3b3c2269de33.png)
 

[Ethers apparently already hides how to discover the RPC endpoint in use](https://github.com/ethers-io/ethers.js/issues/2611), so it would be great if hardhat-ethers didn't further compound the obfuscation ;-)  Apologies in advance if I'm missing something.

**Comment by alcuadrado** [comment]
Hey @aspiers,

> seems to always return { url: 'http://localhost:8545' }

This is unintentional and sounds like a bug.

My recommended way of doing it would be `hre.network.config.url` instead. Note that it will be `undefined` when using the in-process Hardhat Network.

**Reply by fvictorio** [reply]
We are tracking this [here](https://github.com/NomicFoundation/hardhat/issues/1138).
